cli/lsp/diagnostics.rs

```diff
@@ -25,8 +35,299 @@ pub enum DiagnosticSource {
+        disturbed = true;
+      }
+    }
+    Err(err) => {
+      error!("Internal error: {}", err);
+    }
+    _ => (),
+  }
+  performance.measure(mark);
+
+  if disturbed {
+    publish_diagnostics(client, collection, snapshot, performance).await
+  }
+}
+
+fn handle_request(
+  maybe_request: Option<DiagnosticRequest>,
+  collection: &mut DiagnosticCollection,
+  maybe_snapshot: &mut Option<StateSnapshot>,
+) -> bool {
+  match maybe_request {
+    Some(request) => {
+      match request {
+        DiagnosticRequest::Get(specifier, source, tx) => {
+          let diagnostics = collection
+            .diagnostics_for(&specifier, &source)
+            .cloned()
+            .collect();
+          if tx.send(diagnostics).is_err() {
+            error!("DiagnosticServer unable to send response on channel.");
+          }
+        }
+        DiagnosticRequest::Invalidate(specifier) => {
+          collection.invalidate(&specifier)
+        }
+        DiagnosticRequest::Update(snapshot) => *maybe_snapshot = Some(snapshot),
+      }
+      true
+    }
+    _ => false,
+  }
+}
+
+/// A server which calculates diagnostics in its own thread and publishes them
+/// to an LSP client.
+#[derive(Debug)]
+pub struct DiagnosticsServer(mpsc::UnboundedSender<DiagnosticRequest>);
+
+impl DiagnosticsServer {
+  pub fn new(
+    client: Client,
+    ts_server: Arc<tsc::TsServer>,
+    performance: Performance,
+  ) -> Self {
+    let (tx, mut rx) = mpsc::unbounded_channel::<DiagnosticRequest>();
+    let _join_handle = thread::spawn(move || {
+      let runtime = create_basic_runtime();
+      let mut collection = DiagnosticCollection::default();
+
+      runtime.block_on(async {
+        // Some(snapshot) is the flag we use to determine if something has
+        // changed where we will wait for the timeout of changes or a request
+        // that forces us to update diagnostics
+        let mut maybe_snapshot: Option<StateSnapshot> = None;
+
+        loop {
+          let next = rx.recv();
+          tokio::pin!(next);
+
+          let duration = if maybe_snapshot.is_some() {
+            time::Duration::from_millis(DIAGNOSTIC_DEBOUNCE_MS)
+          } else {
+            // we need to await an arbitrary silly amount of time, so this is
+            // 1 year in seconds
+            time::Duration::new(31_622_400, 0)
+          };
+          let debounce = time::sleep(duration);
+          tokio::pin!(debounce);
+
+          // "race" the next message off the rx queue or the debounce timer.
+          // if the next message comes off the queue, the next iteration of the
+          // loop will reset the debounce future.  When the debounce future
+          // occurs, the diagnostics will be updated based on only the most
+          // recent snapshot, thereby "abandoning" any interim states.
+          tokio::select! {
+            _ = &mut debounce => {
+              if let Some(snapshot) = maybe_snapshot {
```

Comment:
I wonder if it might be more efficient to change `maybe_snapshot` into a simple boolean `dirty` or something.
If you find the dirty flag set at this point, you `.lock().unwrap().snapshot()` the `LanguageServer` object right here.

This would ensure that `update_diagnostics()` always operates on the the latest state while avoiding making countless snapshots only to discard them later.

The only reason I can think of that would make this work less well is if the main thread locks the `LanguageServer` object for long stretches (because then the diagnostics thread would have to wait for the lock). BTW it is also possible that this works in our advantage; when the `LanguageServer` is locked by the main thread, presumably it is getting mutated so it might be *good* to wait for it to calm down before kicking off the typescript compiler.

Reply:
Yeah, that reflects what you and I were talking about 1-2-1, so I will re-work it, it makes sense and it is cleaner, and requires a lot less snapshotting.  The locks should be really sort, and we can't really get into a dead-lock with this, so it is fine to wait a bit.